core/base/segmented_range.hpp

```diff
@@ -0,0 +1,372 @@
+        : segmented_range{ptrs, ptrs + 1, num_segments}
+    {}
+
+    /**
+     * Returns the segment at a given index.
+     *
+     * @param segment  the index to access. It must be in `[0, num_segments())`.
+     * @return  the segment at this index.
+     */
+    constexpr segment_type operator[](index_type segment) const
+    {
+        assert(segment >= 0);
+        assert(segment < num_segments());
+        return (*iterator{*this, segment}).segment;
+    }
+
+    /** @return the number of segments in this range. */
+    constexpr index_type num_segments() const { return num_segments_; }
+
+    /** @return an index range representing all segment indices. */
+    constexpr irange<index_type> segment_indices() const
+    {
+        return irange<index_type>{num_segments()};
+    }
+
+    /** @return iterator pointing to the first segment. */
+    constexpr iterator begin() const { return iterator{*this, 0}; }
+
+    /** @return iterator pointing one past the last segment. */
+    constexpr iterator end() const { return iterator{*this, num_segments()}; }
+
+    /** @return iterator pointing to the first segment. */
+    constexpr const index_type* begin_indices() const { return begins_; }
+
+    /** @return iterator pointing one past the last segment. */
+    constexpr const index_type* end_indices() const { return ends_; }
+
+    /** @return the beginning index of the given segment. */
+    constexpr index_type begin_index(index_type segment) const
+    {
+        assert(segment >= 0);
+        assert(segment < num_segments());
+        return begin_indices()[segment];
+    }
+
+    /** @return the end index of the given segment. */
+    constexpr index_type end_index(index_type segment) const
+    {
+        assert(segment >= 0);
+        assert(segment < num_segments());
+        return end_indices()[segment];
+    }
+
+    /** Compares two ranges for equality. */
+    constexpr friend bool operator==(segmented_range lhs, segmented_range rhs)
+    {
+        return lhs.begin_indices() == rhs.begin_indices() &&
+               lhs.end_indices() == rhs.end_indices() &&
+               lhs.num_segments() == rhs.num_segments();
+    }
+
+    /** Compares two ranges for inequality. */
+    constexpr friend bool operator!=(segmented_range lhs, segmented_range rhs)
+    {
+        return !(lhs == rhs);
+    }
+
+private:
+    const index_type* begins_;
+    const index_type* ends_;
+    index_type num_segments_;
+};
+
+
+/**
+ * Represents a range of indices that is segmented into contiguous segments,
+ * mapped into a value array. Each segment has the shape `[begin, end)`, i.e. it
+ * is a half-open interval and points to corresponding entries of the value
+ * array.
+ *
+ * @tparam IndexType  the type of indices used to represent the segments.
+ * @tparam ValueIterator  the iterator type pointing to the values.
+ */
+template <typename IndexType, typename ValueIterator>
+class segmented_value_range {
```

Comment:
I'm wondering a bit how this will work out with the `device_segmented_array` and device implementations of our classes in general. 
In principle, `device_segmented_array` could be replaced directly by this. One thing I would not like about that is that the `device_segmented_array` basically consists only of data members, and in general I would prefer our device implementations to not have member functions. (My reason for that is that I want to generalize the device implementations so that they can be directly exposed to kokkos or similar.)
So right now, I would imagine conversion functions and having some redundancy, instead of replacing `device_segmented_array`.

Reply:
I think that is a heavy restriction. Is there a reason why having no member functions is necessary to expose these directly to kokkos ?

Reply:
It's not necessary. But I would prefer to not provide methods, as I'm not sure if they would also be available in Kokkos code. I guess they would be, since this can run on the GPU, but I would still be a bit hesitant.